src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,281 @@
+      commitsToExclude.map((sha) => '^' + sha)
+    )
+  )
+  const commitsToInclude = rawResponse.split('\n').filter((c) => c !== '')
+
+  return commitsToInclude
+}
+
+// need to transform into async here.
+const generatePackFilesForCommits = (log: Logger, commits: string[]) => {
+  if (commits.length <= 0) {
+    return []
+  }
+
+  const generatePackfiles = (packfilePath: string) => {
+    const packObjectResults = child_process
+      .execSync(`git pack-objects --compression=9 --max-pack-size=3m ${packfilePath}`, {
+        input: commits.join('\n'),
+      })
+      .toString()
+      .split('\n')
+      .filter((sha) => sha)
+      .map((sha) => `${packfilePath}-${sha}.pack`)
+
+    return packObjectResults
+  }
+
+  // Try using tmp folder first:
+  try {
+    const tmpFolder = os.tmpdir()
+    const randomPrefix = String(Math.floor(Math.random() * 10000))
+    const tmpPath = path.join(tmpFolder, randomPrefix)
+
+    return generatePackfiles(tmpPath)
+  } catch (err) {
+    /**
+     * The generation of pack files in the temporary folder (from `os.tmpdir()`)
+     * sometimes fails in certain CI setups with the error message
+     * `unable to rename temporary pack file: Invalid cross-device link`.
+     * The reason why is unclear.
+     *
+     * A workaround is to attempt to generate the pack files in `process.cwd()`.
+     * While this works most of the times, it's not ideal since it affects the git status.
+     * This workaround is intended to be temporary.
+     *
+     * TODO: fix issue and remove workaround.
+     */
+    log.warn(`Failed generation of packfiles in tmpdir: ${err}`)
+    log.warn(`Generating them in ${process.cwd()} instead`)
+    const randomPrefix = String(Math.floor(Math.random() * 10000))
+    const cwdPath = path.join(process.cwd(), randomPrefix)
+
+    return generatePackfiles(cwdPath)
+  }
+}
+
+export const uploadPackfiles = async (
+  log: Logger,
+  request: RequestBuilder,
+  repoURL: string,
+  headCommit: string,
+  packfilePaths: string[]
+) => {
+  // this loop makes sure requests are performed sequentially
+  for (const pack of packfilePaths) {
+    await uploadPackfile(log, request, repoURL, headCommit, pack)
+  }
+}
+
+export const uploadPackfile = async (
+  log: Logger,
+  request: RequestBuilder,
+  repoURL: string,
+  headCommit: string,
+  packfilePath: string
+) => {
+  const pushedSha = JSON.stringify({
+    data: {
+      id: headCommit,
+      type: 'commit',
+    },
+    meta: {
+      repository_url: repoURL,
+    },
+  })
+
+  const form = new FormData()
+
+  form.append('pushedSha', pushedSha, {contentType: 'application/json'})
+  const packFileContent = fs.readFileSync(packfilePath)
+  // The original filename includes a random prefix, so we remove it here
+  const [, filename] = path.basename(packfilePath).split('-')
+  form.append('packfile', packFileContent, {
+    filename,
+    contentType: 'application/octet-stream',
+  })
+
+  return runRequest(log, 'pakfile', () =>
```

Comment:
Will typescript do the right thing here? Or will your function now return a `Promise<Promise<AxiosResponse<T>>>`?

Reply:
Yes, if what you meant is you'd expect this line to be `return await runRequest...`, eslint displays an error `Redundant use of await on a return value.`. The response here will be `Promise<AxiosResponse<T>>`. 

(actually by inspecting the type my editor says it's `Promise<AxiosResponse<AxiosResponse>>` due to a "mistake" in the definition of the type `RequestBuilder`, I just fixed than in another commit).

Reply:
Magic!